src/ripple/shamap/impl/SHAMap.cpp

```diff
@@ -615,68 +803,97 @@ bool SHAMap::delItem (uint256 const& id)
-            int bc = node->getBranchCount ();
-
-            if (bc == 0)
+            int bc = node->getBranchCount();
+            if (is_v2())
             {
-                // no children below this branch
-                prevHash.zero();
-                prevNode.reset ();
-            }
-            else if (bc == 1)
-            {
-                // If there's only one item, pull up on the thread
-                std::shared_ptr<SHAMapItem const> item = onlyBelow (node.get ());
-
-                if (item)
+                assert(bc != 0);
+                if (bc == 1)
                 {
                     for (int i = 0; i < 16; ++i)
                     {
                         if (!node->isEmptyBranch (i))
                         {
-                            node->setChild (i, nullptr);
+                            prevNode = descendThrow(node, i);
                             break;
                         }
                     }
-                    prevNode = std::make_shared<SHAMapTreeNode>(item, type, node->getSeq());
-                    prevHash = prevNode->getNodeHash();
                 }
-                else
+                else  // bc >= 2
                 {
-                    prevHash = node->getNodeHash ();
-                    prevNode = std::move (node);
+                    // This node is now the end of the branch
+                    prevNode = std::move(node);
                 }
             }
             else
             {
-                // This node is now the end of the branch
-                prevHash = node->getNodeHash ();
-                prevNode = std::move (node);
+                if (bc == 0)
+                {
+                    // no children below this branch
+                    prevNode.reset ();
+                }
+                else if (bc == 1)
+                {
+                    // If there's only one item, pull up on the thread
+                    std::shared_ptr<SHAMapItem const> item = onlyBelow (node.get ());
+
+                    if (item)
+                    {
+                        for (int i = 0; i < 16; ++i)
+                        {
+                            if (!node->isEmptyBranch (i))
+                            {
+                                node->setChild (i, nullptr);
+                                break;
+                            }
+                        }
+                        prevNode = std::make_shared<SHAMapTreeNode>(item, type, node->getSeq());
+                    }
+                    else
+                    {
+                        prevNode = std::move (node);
+                    }
+                }
+                else
+                {
+                    // This node is now the end of the branch
+                    prevNode = std::move (node);
+                }
             }
         }
     }
 
     return true;
 }
 
+static
+uint256
+prefix(unsigned depth, uint256 const& key)
+{
+    uint256 r{};
```

Comment:
32 bytes is short enough that I think this is faster.  Though it won't make a whole of difference either way.